astronomer/providers/google/cloud/gke_utils.py

```diff
@@ -0,0 +1,154 @@
+from __future__ import annotations
+
+import logging
+import os
+import tempfile
+from contextlib import contextmanager
+from typing import Generator, Sequence
+
+import yaml
+from airflow.exceptions import AirflowException
+from airflow.providers.google.common.hooks.base_google import GoogleBaseHook
+from airflow.utils.process_utils import execute_in_subprocess, patch_environ
+from google.auth import impersonated_credentials
+from google.cloud.container_v1 import ClusterManagerClient
+from google.oauth2.service_account import Credentials
+
+KUBE_CONFIG_ENV_VAR = "KUBECONFIG"
+
+
+def _get_impersonation_token(
+    creds: Credentials,
+    impersonation_account: str,
+    project_id: str,
+    location: str,
+    cluster_name: str | None = None,
+) -> str:
+    impersonated_creds = impersonated_credentials.Credentials(
+        source_credentials=creds,
+        target_principal=impersonation_account,
+        target_scopes=["https://www.googleapis.com/auth/cloud-platform"],
+    )
+    try:
+        client = ClusterManagerClient(credentials=impersonated_creds)
+        name = f"projects/{project_id}/locations/{location}/clusters/{cluster_name}"
+        client.get_cluster(name=name)
+    except Exception as e:
+        raise Exception(f"Error while creating impersonated creds: {e}") from e
+    return impersonated_creds.token  # type: ignore[no-any-return]
+
+
+def _write_token_into_config(config_name: str, token: str, cluster_context: str | None = None) -> None:
+    # breakpoint()
+    with open(config_name) as input_file:
+        config_content = yaml.safe_load(input_file.read())
+
+    try:
+        if not cluster_context:
+            cluster_context = config_content["current-context"]
+
+        user_name = ""
+        for context in config_content["contexts"]:
+            if context["name"] == cluster_context:
+                user_name = context["context"]["user"]
+                break
+
+        user_index = 0
+        for u_index, user in enumerate(config_content["users"]):
+            if user["name"] == user_name:
+                user_index = u_index
+                break
+
+        # update the token
+        config_content["users"][user_index]["user"]["auth-provider"] = {
+            "name": "gcp",
+            "config": {"access-token": token},
+        }
+    except KeyError as e:
+        logging.exception(e)
+        return
```

Comment:
Do we need to reraise the exception here?

Reply:
This is removed as I rewrite this section with `kubernetes_asyncio.config.kube_config.KubeConfigLoader`